crates/fj-host/src/host.rs

```diff
@@ -1,31 +1,119 @@
-use crossbeam_channel::Receiver;
+use std::thread;
 
-use crate::{Error, Evaluator, Model, ModelEvent, Watcher};
+use crossbeam_channel::{unbounded, Receiver, Sender};
+use fj_operations::shape_processor::ShapeProcessor;
+use winit::event_loop::EventLoopProxy;
 
-/// A Fornjot model host
+use crate::{HostCommand, HostHandle, Model, ModelEvent, Watcher};
+
+// Use a zero-sized error type to silence `#[warn(clippy::result_large_err)]`.
+// The only error from `EventLoopProxy::send_event` is `EventLoopClosed<T>`,
+// so we don't need the actual value. We just need to know there was an error.
+/// Error type for sending on a channel with a closed event loop
+pub struct EventLoopClosed;
+
+/// A Fornjot model host that runs in the background
 pub struct Host {
-    evaluator: Evaluator,
-    _watcher: Watcher,
+    shape_processor: ShapeProcessor,
+    event_loop_proxy: EventLoopProxy<ModelEvent>,
+    command_tx: Sender<HostCommand>,
+    command_rx: Receiver<HostCommand>,
 }
 
 impl Host {
-    /// Create a new instance of `Host`
-    ///
-    /// This is only useful, if you want to continuously watch the model for
-    /// changes. If you don't, just keep using `Model`.
-    pub fn from_model(model: Model) -> Result<Self, Error> {
-        let watch_path = model.watch_path();
-        let evaluator = Evaluator::from_model(model);
-        let watcher = Watcher::watch_model(watch_path, &evaluator)?;
-
-        Ok(Self {
-            evaluator,
-            _watcher: watcher,
-        })
+    /// Create a new `Host` that will process models for an event loop.
+    pub fn new(
+        shape_processor: ShapeProcessor,
+        event_loop_proxy: EventLoopProxy<ModelEvent>,
+    ) -> Self {
+        let (command_tx, command_rx) = unbounded();
+        Self {
+            shape_processor,
+            event_loop_proxy,
+            command_tx,
+            command_rx,
+        }
+    }
+
+    /// Run a background thread to watch for updates and process a model.
+    pub fn spawn(mut self) -> HostHandle {
```

Comment:
Seems a bit superfluous, to have a separate `spawn` method. Maybe it would make for a nicer API to merge `spawn` into `new` rename `HostHandle` to `Host`, and possibly have a `HostInternal` (or whatever; I'm not married to that name) for those private methods.

Reply:
Something like that crossed my mind but for some reason I was stuck to the idea that the `host` variable needed to live outside the event loop.  I implemented your suggestion and it seems a lot cleaner.  It was a fairly big commit so it will need some more review before merging.